File: backend/models/game.py
from pydantic import BaseModel


class MatchingPair(BaseModel):
    chinese: str
    pinyin: str
    english: str
    audio_path: str | None = None


class MatchingRound(BaseModel):
    pairs: list[MatchingPair]


class MadLibsRound(BaseModel):
    sentence_zh: str  # sentence with ____ blank
    sentence_en: str
    pinyin_sentence: str
    vocab_word: str  # correct answer
    options: list[str]  # 4 choices including correct
    rate_limited: bool = False


class MadLibsGenerateReq(BaseModel):
    hsk_level: int


class ScramblerRound(BaseModel):
    sentence_en: str
    words: list[str]           # Scrambled Chinese word tiles (no punctuation)
    correct_order: list[str]   # Correct sequence of words
    full_sentence_zh: str      # Original sentence with punctuation (shown after)
    pinyin_sentence: str


class SentenceCount(BaseModel):
    hsk_level: int
    count: int


class TuneInRound(BaseModel):
    audio_path: str
    correct: str
    correct_pinyin: str
    correct_english: str
    options: list[str]


class AudioCardCount(BaseModel):
    count: int


class DededeRound(BaseModel):
    sentence: str              # sentence with ____ blank
    english: str
    pinyin: str
    answer: str                # correct: 的, 得, or 地
    audio_path: str | None = None  # TTS audio for the correct particle


class ScrambleHarderRound(BaseModel):
    direction: str               # "zh" or "en" — language to unscramble
    prompt: str                  # the other language shown as the clue
    words: list[str]             # all tiles (correct + distractors), shuffled
    correct_order: list[str]     # correct sequence of words to place
    num_correct: int             # how many tiles belong to the answer
    full_sentence_zh: str
    full_sentence_en: str
    pinyin_sentence: str

File: frontend/src/types/game.ts
export interface MatchingPair {
  chinese: string;
  pinyin: string;
  english: string;
  audio_path: string | null;
}

export interface MatchingRound {
  pairs: MatchingPair[];
}

export interface MadLibsRound {
  sentence_zh: string;
  sentence_en: string;
  pinyin_sentence: string;
  vocab_word: string;
  options: string[];
  rate_limited: boolean;
}

export interface ScramblerRound {
  sentence_en: string;
  words: string[];
  correct_order: string[];
  full_sentence_zh: string;
  pinyin_sentence: string;
}

export interface SentenceCount {
  hsk_level: number;
  count: number;
}

export interface TuneInRound {
  audio_path: string;
  correct: string;
  correct_pinyin: string;
  correct_english: string;
  options: string[];
}

export interface AudioCardCount {
  count: number;
}

export interface ScrambleHarderRound {
  direction: string;         // "zh" or "en"
  prompt: string;
  words: string[];
  correct_order: string[];
  num_correct: number;
  full_sentence_zh: string;
  full_sentence_en: string;
  pinyin_sentence: string;
}

export interface DededeRound {
  sentence: string;
  english: string;
  pinyin: string;
  answer: string;   // "的", "得", or "地"
  audio_path: string | null;
}

export type GameType = "matching" | "madlibs" | "scrambler" | "tunein" | "scrambleharder" | "dedede" | "random";
